src/core/varlink-unit.c

```diff
@@ -0,0 +1,423 @@
+                        SD_JSON_BUILD_PAIR_BOOLEAN("CanStop", unit_can_stop_refuse_manual(u)),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("CanReload", unit_can_reload(u)),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("CanIsolate", unit_can_isolate_refuse_manual(u)),
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("CanClean", can_clean_build_json, u),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("CanFreeze", unit_can_freeze(u)),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("CanLiveMount",  unit_can_live_mount(u, /* error= */ NULL) >= 0),
+                        JSON_BUILD_PAIR_UNSIGNED_NON_ZERO("JobId", u->job ? u->job->id : 0),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("NeedDaemonReload", unit_need_daemon_reload(u)),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("ConditionResult", u->condition_result),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("AssertResult", u->assert_result),
+                        JSON_BUILD_PAIR_DUAL_TIMESTAMP_NON_NULL("ConditionTimestamp", &u->condition_timestamp),
+                        JSON_BUILD_PAIR_DUAL_TIMESTAMP_NON_NULL("AssertTimestamp", &u->assert_timestamp),
+                        SD_JSON_BUILD_PAIR_CONDITION(!sd_id128_is_null(u->invocation_id), "InvocationID", SD_JSON_BUILD_UUID(u->invocation_id)),
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("Markers", markers_build_json, &u->markers),
+                        JSON_BUILD_PAIR_CALLBACK_NON_NULL("ActivationDetails", activation_details_build_json, u->activation_details),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("DebugInvocation", u->debug_invocation));
+}
+
+static int list_unit_one(sd_varlink *link, Unit *unit, bool more) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        int r;
+
+        assert(link);
+        assert(unit);
+
+        r = sd_json_buildo(
+                &v,
+                SD_JSON_BUILD_PAIR_CALLBACK("context", unit_context_build_json, unit),
+                SD_JSON_BUILD_PAIR_CALLBACK("runtime", unit_runtime_build_json, unit));
+        if (r < 0)
+                return r;
+
+        if (more)
+                return sd_varlink_notify(link, v);
+
+        return sd_varlink_reply(link, v);
+}
+
+static int lookup_unit_by_pidref(sd_varlink *link, Manager *manager, PidRef *pidref, Unit **ret_unit) {
+        _cleanup_(pidref_done) PidRef peer = PIDREF_NULL;
+        Unit *unit;
+        int r;
+
+        assert(link);
+        assert(manager);
+        assert(ret_unit);
+
+        if (pidref_is_automatic(pidref)) {
+                r = varlink_get_peer_pidref(link, &peer);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to get peer pidref: %m");
+
+                pidref = &peer;
+        } else if (!pidref_is_set(pidref))
+                return -EINVAL;
+
+        unit = manager_get_unit_by_pidref(manager, pidref);
+        if (!unit)
+                return -ESRCH;
```

Comment:
Is ESRCH really the right errno here? It's not distinguishable from exited peer process. ENOENT?